homeassistant/components/recorder/migration.py

```diff
@@ -38,23 +38,31 @@ def migrate_schema(instance):
 
 def _apply_update(engine, new_version):
     """Perform operations to bring schema up to date."""
-    from sqlalchemy import Table
+    from sqlalchemy import Table, Index
     from . import models
 
-    if new_version == 1:
-        def create_index(table_name, column_name):
-            """Create an index for the specified table and column."""
-            table = Table(table_name, models.Base.metadata)
-            name = "_".join(("ix", table_name, column_name))
-            # Look up the index object that was created from the models
-            index = next(idx for idx in table.indexes if idx.name == name)
-            _LOGGER.debug("Creating index for table %s column %s",
-                          table_name, column_name)
-            index.create(engine)
-            _LOGGER.debug("Index creation done for table %s column %s",
-                          table_name, column_name)
+    def create_index(table_name, column_name):
+        """Create an index for the specified table and column."""
+        table = Table(table_name, models.Base.metadata)
+        name = "_".join(("ix", table_name, column_name))
+        # Look up the index object that was created from the models
+        index = next(idx for idx in table.indexes if idx.name == name)
+        _LOGGER.debug("Creating index for table %s column %s",
+                      table_name, column_name)
+        index.create(engine)
+        _LOGGER.debug("Index creation done for table %s column %s",
+                      table_name, column_name)
 
+    if new_version == 1:
         create_index("events", "time_fired")
+    elif new_version == 2:
+        # Create compound start/end index for recorder_runs
+        table = Table("recorder_runs", models.Base.metadata)
+        index = Index("ix_start_end", table.c.start, table.c.end)
+        index.create(engine)
+        # Create indexes for states
+        create_index("states", "last_updated")
+        create_index("states", "created")
```

Comment:
The only thing `create_index` does that doesn't apply to a multi-column index is it generates the name of the index based on the table and column name. If you separate the logic out below the `name = ` line, you should be able to use it for a multi-column index.